challenge-manager/challenges.go

```diff
@@ -5,22 +5,59 @@ package challengemanager
 
 import (
 	"context"
+	"crypto/rand"
 	"fmt"
+	"math/big"
+	"time"
 
-	"github.com/offchainlabs/bold/containers/option"
-	l2stateprovider "github.com/offchainlabs/bold/layer2-state-provider"
 	"github.com/ethereum/go-ethereum/accounts/abi/bind"
 	"github.com/ethereum/go-ethereum/log"
+	"github.com/offchainlabs/bold/containers/option"
+	l2stateprovider "github.com/offchainlabs/bold/layer2-state-provider"
 
 	protocol "github.com/offchainlabs/bold/chain-abstraction"
 	edgetracker "github.com/offchainlabs/bold/challenge-manager/edge-tracker"
 	"github.com/offchainlabs/bold/containers"
 	"github.com/pkg/errors"
 )
 
-// ChallengeAssertion initiates a challenge on an assertion added to the protocol by finding its parent assertion
-// and starting a challenge transaction. If the challenge creation is successful, we add a leaf
-// with an associated history commitment to it and spawn a challenge tracker in the background.
+// HandleCorrectRival is called when the assertion manager has posted a correct
+// rival assertion on the chain and the chllenge manager needs to create a
+// challenge committing to the correct assertion to rival one or more incorrect
+// assertions.
+func (m *Manager) HandleCorrectRival(ctx context.Context, riv protocol.AssertionHash) error {
+	// Generate a random integer between 0 and max delay seconds to wait before
+	// challenging.
+	// This is to avoid all validators challenging at the same time.
+	mds := 1 // default max delay seconds to 1 to avoid panic
+	if m.MaxDelaySeconds() > 1 {
+		mds = m.MaxDelaySeconds()
+	}
+	randSecs, err := randUint64(uint64(mds))
+	if err != nil {
+		return err
+	}
+	time.Sleep(time.Duration(randSecs) * time.Second)
```

Comment:
Maybe we move this to the actual `transact` method of the chain abstraction, as it should be relevant to all moves a staker makes, not just assertion posting

Reply:
I'm actually growing dubious of the utility of this random delay. Here are several thoughts I've had.

1. We don't actually even set `maxDelaySeconds`, so it's going to randomly be 0 or 1 seconds. As soon as we have multiple validators, half of them are going to delay 0 seconds, and half are going to delay 1 second. So, assuming that they would otherwise all be challenging at the same moment, they would still compete with half the honest validators.

2. Since we are moving away from the block subscription model to a model where we try to discover blocks every `avgBlockCreationTime` seconds, we actually have some random variability among the set of honest validators already. That's because they won't all be started at exactly the same second. So, in the example of Arbitrum 1, the 12-second loops are not likely to coincide with one another.

3. Looking at some of the other uses of the `transact` function in the sol-implementation, for example the fast confirmations, it seems like it might actually be counter-productive to insert a delay because the idea is to be "fast."

Given all of the above, I'm sort of inclined to drop the whole idea of a random delay until we see that it really is causing a problem for a set of honest validators. Thoughts?

Reply:
Discussed offline to remove random delay